ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/ForeignKeyTableTableJoin.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.plan;
+
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.google.common.collect.ImmutableList;
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.name.ColumnName;
+import java.util.List;
+import java.util.Objects;
+
+@Immutable
+public class ForeignKeyTableTableJoin<KLeftT, KRightT>
+    implements ExecutionStep<KTableHolder<KLeftT>> {
+
+  private final ExecutionStepPropertiesV1 properties;
+  private final JoinType joinType;
+  private final ColumnName leftJoinColumnName;
+  private final ExecutionStep<KTableHolder<KLeftT>> leftSource;
+  private final ExecutionStep<KTableHolder<KRightT>> rightSource;
+
+  public ForeignKeyTableTableJoin(
+      final ExecutionStepPropertiesV1 props,
+      final JoinType joinType,
+      final ColumnName leftJoinColumnName,
+      final ExecutionStep<KTableHolder<KLeftT>> leftSource,
+      final ExecutionStep<KTableHolder<KRightT>> rightSource
+  ) {
```

Comment:
> I'm not sure if we should have two classes for each time of join. They have repeated code except the extra left column name.

I think it makes sense to have a separate execution step for FK joins since the implementation is different. Execution steps are meant to be fine-grained so we can evolve them separately as needed.

> Also, I'm not sure if the TableTableJoin info is written in the plan. I see that class has a @JsonCreator and it is also referenced in PlanSummary.java as .put(TableTableJoin.class, "JOIN"). Did you verify this part?

PlanSummary.java, ExecutionStep.java, and StepSchemaResolver.java all need to be updated to include the new execution step. Let's add these changes into this PR so we don't hit confusing failures down the line when wiring up the functionality.

> The TableTableJoin accepts an optional keyColumn. Perhaps we can add another parameter to accept an optional nonKeyColumn too?

Unless I'm mistaken, the optional keyColumn in TableTableJoin specifies the name of the _output_ key column. That is not necessary for FK joins because the output key always matches the left source key.